Add E2E tests for the integration in Product Block Editor

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

- Upgrade `@playwright/test` to `^1.42.0` for having the new APIs:
   - https://playwright.dev/docs/api/class-page#page-unroute-all
   - https://playwright.dev/docs/api/class-locator#locator-press-sequentially
- Convert space indentations in `test-snippets.php` to tabs.
- Adjust the product data in `tests/e2e/config/default.json` to fit with WC v3 REST API and remove unused parts.
- Add new E2E API utils for creating variable and variation products.
- Add E2E tests

#### 📌 Checklist (@eason9487)

- [x] Revert 77022c71d358d87a942112c13f9c36b6dca25b33 before merging this PR

### Detailed test instructions:

1. `npm install` to update `@playwright/test` package.
2. `npx playwright install chromium` to update the PlayWright browser as well.
3. `npm run dev` to build files
4. `npm run test:e2e` to see if the E2E test in headless mode can pass.
5. (Optional) `npm run test:e2e -- --ui` to inspect the E2E test in UI mode.
6. View the E2E test result on GitHub Actions: https://github.com/woocommerce/google-listings-and-ads/actions/runs/8376570038/job/22936559491?pr=2327

### Changelog entry

eason9487 committed:
Product block editor E2E test: Prompt to Get Started when not yet finished onboarding

eason9487 committed:
Convert space indentations in test-snippets.php to tabs.

eason9487 committed:
Product block editor E2E test: Check existence and availability of fields for simple product

eason9487 committed:
Adjust the product data in tests/e2e/config/default.json to fit with WC v3 REST API and remove unused parts.

eason9487 committed:
Add new E2E API utils for creating variable and variation products.

eason9487 committed:
Product block editor E2E test: Hide plugin tab for unsupported product types

eason9487 committed:
Product block editor E2E test: Check existence of fields for variable and variation products

eason9487 committed:
Product block editor E2E test: Change channel visibility and check its notice, status, and issues

eason9487 committed:
Product block editor E2E test: Custom block - Product select with text field

eason9487 committed:
Upgrade `@playwright/test` to ^1.42.0

eason9487 committed:
Product block editor E2E test: Generic block transformation - Non-negative integer input field

eason9487 committed:
Product block editor E2E test: Save all product attributes to simple product

eason9487 committed:
Product block editor E2E test: Save product attributes to variation product

eason9487 committed:
Product block editor E2E test: Channel visibility is disabled when hiding in product catalog

eason9487 committed:
Product block editor E2E test: Custom block - Product date and time fields

eason9487 committed:
Temporarily enable pull requests to trigger E2E Tests on GitHub Actions.

eason9487 committed:
Set the tour of product block editor to not display.

puntope commented on .github/workflows/e2e-tests.yml:
```diff
@@ -5,8 +5,6 @@ on:
     branches:
       - "release/**"
   pull_request:
-    branches:
```
Does this mean we will run the e2e tests on each PR?

puntope replied:
Ok, I see you add it for remove it

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+		editorUtils = getProductBlockEditorUtils( page );
+
+		await editorUtils.toggleBlockFeature( true );
+		await api.setOnboardedMerchant();
```
Why have this line here if we are clearing it after?

eason9487 replied:
Apart from the "Prompt to Get Started" case, all other cases need to be tested in the onboarded state.

If it isn't set here, then manual setup will be required when developing other cases, or the "Prompt to Get Started" case will have to be run first.

The main goal is to get the test cases as close to atomicity as possible so that test cases can be tested individually. This makes development and debugging easier and reduces false positives caused by side effects.

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+		);
+
+		// Resume the plugin to onboarded status so that the next test can carry over.
+		await api.setOnboardedMerchant();
```
Based on the comment. Makes more sense for me to add this to the next step.

eason9487 replied:
It's related to https://github.com/woocommerce/google-listings-and-ads/pull/2327#discussion_r1537142428.

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+	} );
+
+	test( 'Check existence and availability of fields for simple product', async () => {
+		await editorUtils.gotoAddProductPage();
```
I guess at this point we are already in ProductPage?  Why not just to change the tab and save a page load?

eason9487 replied:
This is also to make the test case as close as possible to atomicity.

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+		const link = page.getByRole( 'link', { name: 'Get Started' } );
+
+		await expect( link ).toBeVisible();
+		await link.click();
```
I wonder if might be better option to just check the link href and save a page load (for reducing time test time)

eason9487 replied:
Yes, it can be reduced by about 1 ~ 2 seconds. Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2327/commits/caeefeeedff5bb47442cc9ffcb933be98935686b.

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+		await editorUtils.changeToGroupedProduct();
+		await expect( pluginTab ).toHaveCount( 0 );
+
+		await editorUtils.changeToStandardProduct();
+		await expect( pluginTab ).toHaveCount( 1 );
```
This is being tested I guess in line 54 as the default product is standard. So we can omit this

eason9487 replied:
This asserts that the tab can be changed from hidden to visible, which is not equivalent to being visible by default. Also, for the next one assert that the tab goes from visible to hidden.

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+			'This product cannot be shown on any channel because it is hidden from your store catalog.'
+		);
+
+		await editorUtils.clickTab( 'Organization' );
+		await page.getByLabel( 'Hide in product catalog' ).setChecked( false );
+		await editorUtils.save();
+
+		await editorUtils.clickPluginTab();
+		await expect( selection ).toBeEnabled();
+		await expect( help ).toHaveCount( 0 );
```
Why are we testing again this part? Maybe we can reduce execution time as 'Hide in product catalog' is false by default and is tested already.

eason9487 replied:
This asserts that the channel visibility status can be changed back from disabled to enabled.

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+	} );
+
+	test( 'Change channel visibility and check its notice, status, and issues', async () => {
+		await editorUtils.gotoAddProductPage();
+		await editorUtils.fillProductName();
```
I guess these 2 can be omitted and save a page load

eason9487 replied:
This is also to make the test case as close as possible to atomicity.

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+	} );
+
+	test( 'Custom block: Product select with text field', async () => {
+		await editorUtils.gotoAddProductPage();
+		await editorUtils.fillProductName();
```
These 2 can be omitted and save a page load

eason9487 replied:
This is also to make the test case as close as possible to atomicity.

puntope commented on tests/e2e/specs/product-editor/block-integration.test.js:
```diff
@@ -0,0 +1,693 @@
+	} );
+
+	test.afterEach( async () => {
+		await page.unrouteAll();
```
Why we need this?

eason9487 replied:
This is to removes all routes created with `page.route()` via the `editorUtils.mockChannelVisibility` calls in the "Change channel visibility and check its notice, status, and issues" case. In this way, the side effects between test cases can be reduced.

puntope reviewed (commented):
Hi @eason9487 Thanks for your work in E2E test.

All is working fine. Not a strong blocker. However, I worry about the time for the tests to finish execution.

2-3mins

<img width="1406" alt="Screenshot 2024-03-24 at 12 31 15" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/bef3282e-6667-4d8c-b62d-a918cbfd578a">

I left some comments in an attempt to reduce unnecessary saves/loads. I will wait for your feedback before approve.

eason9487 committed:
Product block editor E2E test: Assert the link path instead of the page after redirection to reduce a bit of test execution time.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2327#discussion_r1536657034

eason9487 commented:
Hi @puntope, thanks for the review. Could you take another look?

> However, I worry about the time for the tests to finish execution.

The long execution time is mainly because the Product Block Editor page loading takes 6~10 seconds, which is rather uncontrollable from the plugin side.

Currently, each test has at least one atomicity that can be executed independently and has no side effects. Allowing the next case to continue where the finished states of the previous one would reduce some overall time but at the cost of making development and debugging more difficult. I'm leaning toward keeping test cases as atomic as possible.

puntope reviewed (approved):
Thanks for the clarifications @eason9487 

Sad the Product Editor loads causes those delays. I'm approving it since it doesn't seem to have any action on our side to mitigate it.

eason9487 committed:
Revert "Temporarily enable pull requests to trigger E2E Tests on GitHub Actions."

This reverts commit 77022c71d358d87a942112c13f9c36b6dca25b33.
